src/main/java/org/broadinstitute/hellbender/tools/walkers/variantutils/SelectVariants.java

```diff
@@ -618,51 +674,88 @@ public void apply(VariantContext vc, ReadsContext readsContext, ReferenceContext
-        // If exclude non-variants is called, it is a polymorphic variant, but not a spanning deletion, filtering will not occur
-        // True iff exclude-filtered is not called or the filteredGenotypeToNocall is not already filtered
+        // After subsetting samples (e.g. to a single sample), it's possible that:
+        //     1. none of the remaining samples has the variant in question (i.e. site is no longer polymorphic);
+        //     2. the only remaining alternate allele is spanning deletion (*)
+        // If this is the case, we call it a non-variant and remove it from the output based on {@code excludeNonVariants}
+        final boolean nonVariant = ! result.isPolymorphicInSamples() || GATKVariantContextUtils.isSpanningDeletionOnly(result);
+        if (excludeNonVariants && nonVariant){
+            return;
+        }
 
-        if ((!XLnonVariants || (filteredGenotypeToNocall.isPolymorphicInSamples() && !GATKVariantContextUtils.isSpanningDeletionOnly(filteredGenotypeToNocall)))
-                && (!XLfiltered || !filteredGenotypeToNocall.isFiltered()))
-        {
+        if (!applyJexlFiltersBeforeFilteringGenotypes && ! passesJexlFilters(result)){
+            return;
+        }
 
-            // Write the subsetted variant if it matches all of the expressions
-            boolean failedJexlMatch = false;
+        pendingVariants.add(dropAnnotations(result, infoAnnotationsToDrop, genotypeAnnotationsToDrop));
+    }
 
-            try {
-                for (VariantContextUtils.JexlVCMatchExp jexl : jexls) {
-                    if (invertLogic(!VariantContextUtils.match(filteredGenotypeToNocall, jexl), invertSelect)){
-                        failedJexlMatch = true;
-                        break;
-                    }
+    /**
+     *  Applies JEXL filters
+     *
+     *  - Expressions that contain the logical-and (&&) should appear in a single -select argument.
+     *  - When an annotation is absent, the logical-or returns false, even when the other argument evaluates to true.
+     *   (This behavior may be updated by changing {@code JEXLMap.howToTreatMissingValues})
+     *  - When multiple -select arguments are given, the logical expressions are combined with the logical-or operator. In particular,
+     *    we do not currently support complex logical expressions involving both logical-and's and logical-or's. e.g. (x || y) && z
+     *
+     * When both the INFO (--select) and genotype (--select-genotype) filters are specified, the current behavior is to take the logical-or of them.
+     * We should support combining by logical-and if the users request it.
+     */
+    private boolean passesJexlFilters(final VariantContext vc){
+        if (infoJexls.isEmpty() && genotypeJexls.isEmpty()){
+            return true;
+        }
+
+        try {
+            // ##### Apply INFO JEXL filters #####
+            for (VariantContextUtils.JexlVCMatchExp jexl : infoJexls) {
```

Comment:
Add reminder that the "info" jexls might contain expressions that access the genotypes

Reply:
Done